### DISCUSSION TITLE
Drain Inflow

### DISCUSSION THREAD

**Discussion by cneyens** [MODFLOW-ORG/modflow6#1194] [Ideas]
Hello,

In a recent project, we've been looking for ways to simulate a triggered influx from a control structure, where influx into the groundwater system only occurs when the head in a cell falls below a certain threshold. The inflow depends on some reference head and a conductance. Conceptually, this is somewhat the inverse of a DRAIN feature, where water is removed when the head is above a given level. I've not yet found a way to do this with MODFLOW6. In  the past, we've adapted the source code of MODFLOW to incorporate these kinds of features (e.g. a [maximum-flux constraint for dewatering wells](https://github.com/cneyens/MODFLOW-MAXDRN)), but with our shift to MODFLOW6 and the availability of the MODFLOW-API, this seems a rather cumbersome approach. I was wondering if this is something that you feel is within the scope of the code and could be implemented, or rather to be done with the MODFLOW-API ? Or perhaps I'm overlooking some way to do this with the current MF6 code ? 

Kind regards,
Cas

**Comment by mjr-deltares** [comment]
Hi Cas, to me this sounds like a job for the API. You could write a small python script around the MODFLOW shared library and configure a GHB where you want the influx to happen. Then in your script you decide based on current head values whether the GHB conductance should be set to zero or not. The GHB has observations as well, so you can monitor the process explicitly. This would get you really quickly what you need. I personally do not know about any 'tricks' to configure your feature with existing functionality, but maybe others do.

**Comment by christianlangevin** [comment]
I would second what @mjr-deltares said in his answer.  With the [modflowapi](https://github.com/MODFLOW-USGS/modflowapi) Python Package, it is relatively easy for someone to implement the custom functionality of the MAXDRN version of MODFLOW that you refer to in your message.  In your case, I see two ways for you to do this:

1.  The user would create a model that contains a GHB Package with a list of GHB entries that you may want to modify.  Instead of running the model with the regular `mf6.exe` you would instead run the model using the `libmf6.dll`, which contains the identical routines as `mf6.exe` but is bundled as a library instead.  The [modflowapi](https://github.com/MODFLOW-USGS/modflowapi) Python package is then used to run the simulation using the `llibmf6.dll`.  Running MF6 this way allows you to intervene during this simulation.  There are multiple ways to intervene, but one approach is to use the callback functionality described [here](https://github.com/MODFLOW-USGS/modflowapi/blob/develop/examples/notebooks/MODFLOW-API%20extensions%20objects.ipynb) to modify the conductance value for each GHB entry using the simulated head.  In the section below of the callback function 
```
    if callback_step == Callbacks.iteration_start:
        # we can implement complex solutions to boundary conditions here!
        pass
```
any changes that you make here will update for each MODFLOW iteration.  So here is where you can check the aquifer cell head and then change the conductance value accordingly.  The budget tables, values written to the budget file, etc. will automatically  use these adjusted flows from this calculation.  Note that you may need to store the original conductances in Python somewhere so they can be reset.  Alternatively, maybe a copy of conductance could be specified by the user as an auxiliary variable, and this copy would then be used for each iteration to specify the adjusted conductance (just thinking out loud here).  There are a variety of ways to do this.
2.  @spaulins-usgs is working on flopy-plugins, a new infrastructure for users to write custom packages in flopy and have them available as a Python package that others can install.  We are getting close to releasing a [first draft of this capability](https://github.com/modflowpy/flopy/pull/1746).  One of the first examples of this flopy plugin is a new package that allows the user to assign a different river conductance depending on whether river leakage is positive or negative.

So the bottom line here is that we think the API, in combination with the modflowapi and/or flopy-plugins offer some powerful ways to do what you want.

Lastly, just as a random thought, could you use a trick where you have both a GHB and a DRN Package and have the DRN Package remove the GHB addition above the specified threshhold?  It's a bit hacky, but may work.  Just a thought.

**Comment by cneyens** [comment]
Thank you very much for your input! We'll look into the MODFLOW-API for this.
